Hardware/CMockHardwareMap.cpp

```diff
@@ -0,0 +1,297 @@
+            m_run_period_ms / 1000 / m_heat_capacity[i][0];
+
+        /**
+         * Update temperature of the radiator stage. radiator_flow is heat
+         * flowing from radiator into incubator ambient air. From radiator point
+         * of view this is net loss of heat.
+         * The radiator temperature is then updated based on positive flow from
+         * heater and loss of heat to incubator.
+         * Total radiator heat flow is also accumulated to calculate incubator
+         * temperature. Negative sign is due to radiator loss being net positive
+         * flow for incubator.
+         */
+        radiator_flow = (m_temperature[i][1] - m_incubator_temperature) /
+                        m_heat_conductance[i][1];
+        m_temperature[i][1] += (heater_flow - radiator_flow) * m_run_period_ms /
+                               1000 / m_heat_capacity[i][1];
+        total_radiator_flow -= radiator_flow;
+    }
+    float incubator_flow;
+    incubator_flow =
+        (m_incubator_temperature - m_ambient_temperature) / m_incubator_loss;
+    m_incubator_temperature += (total_radiator_flow - incubator_flow) *
+                               m_run_period_ms / 1000 / m_incubator_capacity;
+}
+
+/**
+ * @note: This method is not for sending commands like >setHeaterPower(). This
+ * method is to simulate input from real world conditions changing as the device
+ * is being tested. For example: induce failure of a heater or thermistor. Or
+ * set different ambient temperature.
+ */
+bool CMockHardwareMap::newCommand(ICommand *p_command,
+                                  IComChannel *p_comchannel)
+{
+    bool b_command_recognised = false;
+    /**
+     * Command to modify mock hardware ambient temperature
+     * >setambient(20);
+     */
+    if (p_command->getName()->compare("setambient"))
+    {
+        float temperature = (*p_command)[0];
+        if ((temperature < MIN_TEMP) || (MAX_TEMP < temperature))
+        {
+            p_comchannel->send("Arguments out of bounds.");
+        }
+        else
+        {
+            m_ambient_temperature = temperature;
+        }
+        b_command_recognised = true;
+    }
+    /**
+     * Command to modify power rating of the heater. Rating refers to heater
+     * full power output. Channel is optional and if omitted command is applied
+     * to all channels.
+     * >setRating(rating, [channel])
+     */
+    if (p_command->getName()->compare("setrating"))
+    {
+        uint8_t channel = (uint8_t)(*p_command)[1];
+        float rating = (*p_command)[0];
+        if ((rating < 0) || (channel > HARD_PWM_OUTPUTS))
+        {
+            p_comchannel->send("Arguments out of bounds.");
+        }
+        else
+        {
+            setrating(rating, channel);
+        }
+        b_command_recognised = true;
+    }
+    /**
+     * Command to modify heat capacity of the heater. Channel is optional and if
+     * omitted command is applied to all channels.
```

Comment:
This is why I think an argument count check could be useful. 
If someone sends a command without the channel argument, this method will try to retrieve a value that doesn't exist (`(uint8_t)(*p_command)[2]`). 

The way my parser handles this is by sending an ERROR_VALUE of 999.9, which is easy to recognise. 
In this case it will only going to send the `"Arguments out of bounds"` error and stop. 

I also think there should be a note clarifying if channels start from 0 or 1.

Reply:
There is a convention that non-existent values should be assigned zero. 999.9 is not a good choice because in some circumstances this might be a valid value.

Reply:
Ok, I didn't know that, I'll fix that in the parser. That's actually the reason I set it to 999.9, because I thought 0 could be an actual value.